### ISSUE TITLE
ICS20 correctly determining source chain

### ISSUE THREAD

**Issue by colin-axner** [cosmos/ibc#454]
Currently the ICS20 spec uses the following to determine the source chain of a token:

```
 // inspect the denomination to determine whether or not we are the source chain
  prefix = "{destPort}/{destChannel}"
  source = denomination.slice(0, len(prefix)) === prefix
```

while this normally works fine, it will fail when two channel ends have the same portID and channelID. It will presume the the token is always source since

`{destPort}/{destChannel} == {sourcePort}/{sourceChannel}`

The sends will simply fail if the user does not have a sufficient balance. It could lead to a user transferring tokens it did not want to. 

I think having a user provided source flag is probably the cleanest option. I think trying to restrict naming requirements unnecessary since this would need to be added to channel handshakes. It'd be strange if an application started imposing restrictions on TAO functionality.

**Comment by colin-axner** [comment]
I think the current code also breaks with overlapping channelID's ie

```
destPort = transfer
destChannel = mychannel

sourcePort = transfer
sourceChannel = mychannelissupercool

prefix = transfer/mychannel
denomination = transfer/mychannelissupercool
source = "transfer/mychannelissupercool".slice(0, len(prefix)) -> "transfer/mychannel"
```

**Comment by cwgoes** [comment]
> while this normally works fine, it will fail when two channel ends have the same portID and channelID

Yes, it will behave in undesirable fashion in that case. Good catch.

> I think having a user provided source flag is probably the cleanest option.

I think so as well.

> I think the current code also breaks with overlapping channelID's ie

Yes, this should also be solved by a manual `source` flag, though.

**Comment by gjermundgaraba** [comment]
This issue was mistakenly reopened during some integration work, so closing again now.
